File: content-hosting/scripts/validate-content.mjs
#!/usr/bin/env node
/**
 * Validates `public/content/` before every `firebase deploy --only hosting` (ADR 0015). Mirrors
 * the same structural rules `ContentValidator`/`ContentImporter` enforce on the Android side
 * (`app/src/main/java/com/sangusantri/app/data/content/ContentValidator.kt`) — this script is
 * intentionally simple tooling, not a second copy of that Kotlin logic; the Android importer
 * remains the authoritative, final validation gate regardless of what this script checks.
 *
 * Usage: node scripts/validate-content.mjs [--previous <path-to-previous-catalog.json>]
 * Exit code 0 = valid, non-zero = at least one error printed to stderr.
 */
import { readFileSync, existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

const SUPPORTED_SCHEMA_VERSION = 1;
const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const contentDir = path.resolve(scriptDir, "..", "public", "content");
const catalogPath = path.join(contentDir, "catalog.json");

const errors = [];

function readJson(filePath, label) {
  if (!existsSync(filePath)) {
    errors.push(`${label}: file not found at ${filePath}`);
    return null;
  }
  const raw = readFileSync(filePath, "utf-8");
  try {
    return JSON.parse(raw);
  } catch (cause) {
    errors.push(`${label}: invalid JSON (${cause.message})`);
    return null;
  }
}

function isBlank(value) {
  return typeof value !== "string" || value.trim().length === 0;
}

function validateCatalog(catalog) {
  if (catalog.schemaVersion !== SUPPORTED_SCHEMA_VERSION) {
    errors.push(`catalog.json: unsupported schemaVersion ${catalog.schemaVersion}`);
  }
  if (!Array.isArray(catalog.items)) {
    errors.push("catalog.json: items must be an array");
    return;
  }

  const seenIds = new Set();
  for (const item of catalog.items) {
    if (isBlank(item.id)) {
      errors.push("catalog.json: item id must not be blank");
      continue;
    }
    if (seenIds.has(item.id)) {
      errors.push(`catalog.json: duplicate item id "${item.id}"`);
    }
    seenIds.add(item.id);

    if (!Number.isInteger(item.version) || item.version <= 0) {
      errors.push(`catalog.json item "${item.id}": version must be a positive integer`);
    }
    if (isBlank(item.title)) errors.push(`catalog.json item "${item.id}": title must not be blank`);
    if (isBlank(item.description)) errors.push(`catalog.json item "${item.id}": description must not be blank`);
    if (isBlank(item.contentUrl)) errors.push(`catalog.json item "${item.id}": contentUrl must not be blank`);
  }
}

/** contentUrl is always rooted at "/content/..." (ADR 0015) — resolve it to a real file on disk. */
function resolveContentFilePath(contentUrl) {
  const relative = contentUrl.replace(/^\/content\//, "");
  return path.join(contentDir, relative);
}

function validateContentFile(item) {
  const filePath = resolveContentFilePath(item.contentUrl);
  const file = readJson(filePath, `content file for "${item.id}" (${item.contentUrl})`);
  if (file === null) return;

  if (file.schemaVersion !== SUPPORTED_SCHEMA_VERSION) {
    errors.push(`${item.contentUrl}: unsupported schemaVersion ${file.schemaVersion}`);
  }
  if (file.id !== item.id) {
    errors.push(`${item.contentUrl}: file id "${file.id}" does not match catalog id "${item.id}"`);
  }
  if (file.version !== item.version) {
    errors.push(
      `${item.contentUrl}: file version ${file.version} does not match catalog version ${item.version}`,
    );
  }
  if (isBlank(file.sourceName)) errors.push(`${item.contentUrl}: sourceName must not be blank`);
  if (isBlank(file.sourceUrl)) errors.push(`${item.contentUrl}: sourceUrl must not be blank`);

  if (!Array.isArray(file.steps) || file.steps.length === 0) {
    errors.push(`${item.contentUrl}: steps must not be empty`);
    return;
  }

  const seenStepIds = new Set();
  for (const step of file.steps) {
    if (isBlank(step.id)) {
      errors.push(`${item.contentUrl}: step id must not be blank`);
    } else if (seenStepIds.has(step.id)) {
      errors.push(`${item.contentUrl}: duplicate step id "${step.id}"`);
    } else {
      seenStepIds.add(step.id);
    }
    if (isBlank(step.arabicText)) errors.push(`${item.contentUrl} step "${step.id}": arabicText must not be blank`);
    if (isBlank(step.translation)) errors.push(`${item.contentUrl} step "${step.id}": translation must not be blank`);
    if (!Number.isInteger(step.repeatTarget) || step.repeatTarget < 1) {
      errors.push(`${item.contentUrl} step "${step.id}": repeatTarget must be at least 1`);
    }
  }
}

function validateNoVersionRegression(catalog, previousCatalogPath) {
  if (!previousCatalogPath) return;
  const previous = readJson(previousCatalogPath, "previous catalog");
  if (previous === null || !Array.isArray(previous.items)) return;

  const previousVersionById = new Map(previous.items.map((item) => [item.id, item.version]));
  for (const item of catalog.items) {
    const previousVersion = previousVersionById.get(item.id);
    if (previousVersion !== undefined && item.version < previousVersion) {
      errors.push(
        `catalog.json item "${item.id}": version ${item.version} regresses from previously deployed ${previousVersion}`,
      );
    }
  }
}

function parseArgs(argv) {
  const previousIndex = argv.indexOf("--previous");
  return {
    previousCatalogPath: previousIndex >= 0 ? argv[previousIndex + 1] : null,
  };
}

function main() {
  const { previousCatalogPath } = parseArgs(process.argv.slice(2));
  const catalog = readJson(catalogPath, "catalog.json");
  if (catalog === null) {
    printResultAndExit();
    return;
  }

  validateCatalog(catalog);
  if (Array.isArray(catalog.items)) {
    for (const item of catalog.items) {
      if (!isBlank(item.id) && !isBlank(item.contentUrl)) {
        validateContentFile(item);
      }
    }
    validateNoVersionRegression(catalog, previousCatalogPath);
  }

  printResultAndExit();
}

function printResultAndExit() {
  if (errors.length === 0) {
    console.log(`content-hosting validation passed (${contentDir})`);
    process.exit(0);
  }
  console.error(`content-hosting validation failed with ${errors.length} error(s):`);
  for (const error of errors) {
    console.error(`  - ${error}`);
  }
  process.exit(1);
}

main();

File: tools/ci/validate_content.py
#!/usr/bin/env python3
"""Pre-merge check for the bundled content tree (ADR 0015, docs/operations/CONTENT_GOVERNANCE.md).

Mirrors the rules `data/content/ContentValidator.kt` and `data/content/ContentImporter.kt` apply on
device. The app fails closed on a bad catalog, so a mistake here does not put wrong text in front of
a reader — it makes content silently stop updating instead. Catching it on the pull request is the
difference between a red check and a quiet outage nobody notices for a week.

Only one tree lives in this repository now:

  app/src/main/assets/content/   bundled baseline, ships inside the APK

What remote sync fetches is no longer a checked-in file tree — it is the CMS API
(`../cms/api`, deployed on Vercel), whose own responses are covered by its Go handler tests and by
`app/src/test/.../CmsApiContractTest.kt`, which runs this same validator's Kotlin twin over
captured API responses.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

SUPPORTED_SCHEMA_VERSION = 1
CONTENT_PATH_PREFIX = "/content/"
HTTPS_SCHEME = "https://"

TREES = (Path("app/src/main/assets/content"),)

errors: list[str] = []


def fail(where: str, message: str) -> None:
    errors.append(f"{where}: {message}")


def is_origin_relative_content_path(url: str) -> bool:
    """Same rule as ContentValidator.isOriginRelativeContentPath.

    An absolute URL here would override Retrofit's base URL outright, and the bundled reader feeds
    the same string to AssetManager.open.
    """
    return (
        url.startswith(CONTENT_PATH_PREFIX)
        and "//" not in url
        and "\\" not in url
        and not any(c.isspace() for c in url)
        and ".." not in url.split("/")
    )


def is_allowed_image_url(url: str | None) -> bool:
    """Same rule as ContentValidator.isAllowedImageUrl."""
    if url is None:
        return True
    return url.startswith(HTTPS_SCHEME) and len(url) > len(HTTPS_SCHEME) and not any(c.isspace() for c in url)


def check_catalog_item(where: str, item: dict, seen_ids: set[str]) -> None:
    item_id = item.get("id", "")
    if not item_id.strip():
        fail(where, "catalog item id must not be blank")
        return
    if item_id in seen_ids:
        fail(where, f"duplicate catalog item id {item_id!r}")
    seen_ids.add(item_id)

    if item.get("version", 0) <= 0:
        fail(where, f"item {item_id}: version must be positive")
    for field in ("title", "description"):
        if not str(item.get(field, "")).strip():
            fail(where, f"item {item_id}: {field} must not be blank")

    content_url = item.get("contentUrl", "")
    if not content_url.strip():
        fail(where, f"item {item_id}: contentUrl must not be blank")
    elif not is_origin_relative_content_path(content_url):
        fail(where, f"item {item_id}: contentUrl {content_url!r} is not an origin-relative path under {CONTENT_PATH_PREFIX}")

    if not is_allowed_image_url(item.get("imageUrl")):
        fail(where, f"item {item_id}: imageUrl {item.get('imageUrl')!r} must be an https URL")


def check_content_file(where: str, item: dict, path: Path) -> None:
    item_id = item.get("id", "?")
    if not path.is_file():
        fail(where, f"item {item_id}: contentUrl points at {path}, which does not exist")
        return

    try:
        content = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        fail(f"{path}", f"not valid UTF-8 JSON: {exc}")
        return

    if content.get("schemaVersion") != SUPPORTED_SCHEMA_VERSION:
        fail(f"{path}", f"unsupported schemaVersion {content.get('schemaVersion')}")

    # ContentImporter rejects a file whose id/version disagree with the catalog entry that named it,
    # so a mismatch here means the item would never import on any device.
    if content.get("id") != item_id:
        fail(f"{path}", f"id {content.get('id')!r} does not match catalog id {item_id!r}")
    if content.get("version") != item.get("version"):
        fail(f"{path}", f"version {content.get('version')} does not match catalog version {item.get('version')}")

    for field in ("sourceName", "sourceUrl"):
        if not str(content.get(field, "")).strip():
            fail(f"{path}", f"{field} must not be blank — attribution is mandatory (CONTENT_GOVERNANCE.md)")

    steps = content.get("steps") or []
    if not steps:
        fail(f"{path}", "steps must not be empty")
    step_ids: set[str] = set()
    for index, step in enumerate(steps):
        label = f"step[{index}] ({step.get('id', '?')})"
        step_id = str(step.get("id", ""))
        if not step_id.strip():
            fail(f"{path}", f"{label}: id must not be blank")
        elif step_id in step_ids:
            fail(f"{path}", f"{label}: duplicate step id")
        step_ids.add(step_id)

        if not str(step.get("arabicText", "")).strip():
            fail(f"{path}", f"{label}: arabicText must not be blank")
        if not str(step.get("translation", "")).strip():
            fail(f"{path}", f"{label}: translation must not be blank")
        if step.get("repeatTarget", 0) < 1:
            fail(f"{path}", f"{label}: repeatTarget must be at least 1")


def check_tree(tree: Path) -> dict | None:
    catalog_path = tree / "catalog.json"
    if not catalog_path.is_file():
        fail(str(tree), "catalog.json is missing")
        return None

    try:
        catalog = json.loads(catalog_path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        fail(str(catalog_path), f"not valid UTF-8 JSON: {exc}")
        return None

    where = str(catalog_path)
    if catalog.get("schemaVersion") != SUPPORTED_SCHEMA_VERSION:
        fail(where, f"unsupported schemaVersion {catalog.get('schemaVersion')}")

    seen_ids: set[str] = set()
    for item in catalog.get("items", []):
        check_catalog_item(where, item, seen_ids)
        content_url = item.get("contentUrl", "")
        if is_origin_relative_content_path(content_url):
            check_content_file(where, item, tree / content_url[len(CONTENT_PATH_PREFIX):])

    return catalog


def main() -> int:
    for tree in TREES:
        check_tree(tree)

    # There used to be a bundled-vs-hosted version parity check here, because remote sync never
    # downgrades: a remote entry older than the bundled one imports on a fresh install and is then
    # permanently skipped by every sync. That hazard is unchanged, but the remote side is now the
    # CMS database rather than a second tree in this repository, so it cannot be checked from a
    # pull request without a network call. Keep bundled asset versions at or below the CMS's.

    if errors:
        for error in errors:
            print(f"::error::{error}")
        print(f"\n{len(errors)} content problem(s) found.", file=sys.stderr)
        return 1

    print(f"Content OK — {len(TREES)} tree(s) validated.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
